### PR TITLE
DNM: Split rgw_admin.cc into several files

### PR DESCRIPTION
**Pull request by cache-nez**

### CODE REVIEW

**Comment by cache-nez** [comment]
@oritwas , for the discussion we had in https://github.com/ceph/ceph/pull/19984:
- commits are now signed,
- no changes to submodules,
- command line parameters parsing is split into several functions,
- my motivation for naming the enum with the commands was enforcing the correctness of values in the code and clarifying what opt_cmd is when passing it to functions. Let's discuss this change if you feel like.
Did I miss anything?

File: src/rgw/CMakeLists.txt [hunk 1]
```diff
@@ -1,20 +1,20 @@
 if(WITH_TESTS)
-add_executable(ceph_rgw_jsonparser
-  rgw_jsonparser.cc)
-target_link_libraries(ceph_rgw_jsonparser
-  rgw_a
-  global)
+  add_executable(ceph_rgw_jsonparser
```
Comment (oritwas) [inline src/rgw/CMakeLists.txt]: please do not change formatting, it makes it harder to review actual code change.
you can do it in  a separate cleanup patch

File: src/rgw/CMakeLists.txt [hunk 2]
```diff
@@ -257,12 +265,12 @@ install(TARGETS rgw DESTINATION ${CMAKE_INSTALL_LIBDIR})
...
     target_include_directories(cephd_rgw_base SYSTEM PUBLIC ${FCGI_INCLUDE_DIR})
   endif()
   set_target_properties(cephd_rgw_base PROPERTIES COMPILE_DEFINITIONS BUILDING_FOR_EMBEDDED)
   merge_static_libraries(cephd_rgw cephd_rgw_base rgw_a radosgw_a)
-  if(WITH_RADOSGW_FCGI_FRONTEND) 
+  if(WITH_RADOSGW_FCGI_FRONTEND)
     target_link_libraries(cephd_rgw ${FCGI_LIBRARY})
   endif()
-endif()
+endif()
```
Comment (oritwas) [inline src/rgw/CMakeLists.txt]: nit: removed newline

File: src/rgw/rgw_admin.cc [hunk 3]
```diff
@@ -1,1011 +1,55 @@
 // -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
 // vim: ts=8 sw=2 smarttab
 
-#include <errno.h>
+#include <cerrno>
```
Comment (oritwas) [inline src/rgw/rgw_admin.cc]: why is this needed?

File: src/rgw/rgw_admin.cc [hunk 4]
```diff
@@ -1,1011 +1,55 @@
...
 #include <iostream>
 #include <sstream>
+#include <cstdlib>
```
Comment (oritwas) [inline src/rgw/rgw_admin.cc]: please use stdlib, but not sure why does you need to add this include

File: src/rgw/rgw_admin.cc [hunk 5]
```diff
@@ -1,1011 +1,55 @@
...
 #include "auth/Crypto.h"
 #include "compressor/Compressor.h"
 
-#include "common/armor.h"
 #include "common/ceph_json.h"
 #include "common/config.h"
-#include "common/ceph_argparse.h"
-#include "common/Formatter.h"
-#include "common/errno.h"
-#include "common/safe_io.h"
```
Comment (oritwas) [inline src/rgw/rgw_admin.cc]: are you sure you don't need common/errno.h?

File: src/rgw/rgw_admin.cc [hunk 6]
```diff
@@ -1,1011 +1,55 @@
...
-#include "rgw_sync_log_trim.h"
 #include "rgw_data_sync.h"
 #include "rgw_rest_conn.h"
 #include "rgw_realm_watcher.h"
 #include "rgw_role.h"
 #include "rgw_reshard.h"
 
-
-#define dout_context g_ceph_context
-#define dout_subsys ceph_subsys_rgw
```
Comment (oritwas) [inline src/rgw/rgw_admin.cc]: please don't remove those, they are needed for logging into rgw log file.

File: src/rgw/rgw_admin.cc [hunk 7]
```diff
@@ -1061,6146 +67,1088 @@ class StoreDestructor {
...
-  if (ret < 0) {
-    cerr << "ERROR: failed to read input: " << cpp_strerror(-ret) << std::endl;
-    return ret;
-  }
-  JSONParser p;
-  if (!p.parse(bl.c_str(), bl.length())) {
-    cout << "failed to parse JSON" << std::endl;
-    return -EINVAL;
-  }
+  boost::intrusive_ptr<CephContext> cct = global_init(nullptr, args, CEPH_ENTITY_TYPE_CLIENT,
```
Comment (oritwas) [inline src/rgw/rgw_admin.cc]: it was auto, don't see the need to change it.
it will make review easier.

File: src/rgw/rgw_admin.cc [hunk 8]
```diff
@@ -1061,6146 +67,1088 @@ class StoreDestructor {
...
 
-  try {
-    t.decode_json(&p, k);
-  } catch (JSONDecoder::err& e) {
-    cout << "failed to decode JSON input: " << e.message << std::endl;
-    return -EINVAL;
-  }
-  return 0;
-}
+  rgw_user user_id;
```
Comment (oritwas) [inline src/rgw/rgw_admin.cc]: two options:
1. create a new struct with those variables and pass it to the parsing functions. 
2. make those variables global and avoid passing them to the paring functions.

improvement for the future is using boost:optional

File: src/rgw/rgw_admin.cc [hunk 9]
```diff
@@ -1061,6146 +67,1088 @@ class StoreDestructor {
...
 
-static int read_current_period_id(RGWRados* store, const std::string& realm_id,
-                                  const std::string& realm_name,
-                                  std::string* period_id)
-{
-  RGWRealm realm(realm_id, realm_name);
-  int ret = realm.init(g_ceph_context, store);
-  if (ret < 0) {
-    std::cerr << "failed to read realm: " << cpp_strerror(-ret) << std::endl;
+  if (raw_storage_op) {
```
Comment (oritwas) [inline src/rgw/rgw_admin.cc]: here using an abstract class and a factory pattern will help remove this huge switch.
see rgw_op.h : you will need a handler class and an op class (in this case admin command class) for each command type (user, period, bucket , ...)

Comment (cache-nez) [reply src/rgw/rgw_admin.cc]: Will look, thanks!

File: src/rgw/rgw_admin.cc (hunk 6 above)
Comment (cache-nez) [reply src/rgw/rgw_admin.cc]: Those are now defined in rgw_admin_common.h.

File: src/rgw/rgw_admin.cc (hunk 3 above)
Comment (cache-nez) [reply src/rgw/rgw_admin.cc]: cerrno is a C++ header while errno.h is a C one. No difference in behavior.

File: src/rgw/rgw_admin.cc (hunk 4 above)
Comment (cache-nez) [reply src/rgw/rgw_admin.cc]: Same as with cerrno vs errno.h

File: src/rgw/rgw_admin.cc (hunk 7 above)
Comment (cache-nez) [reply src/rgw/rgw_admin.cc]: My reasoning was that it's not clear what cct was since the type has never been explicitly mentioned, so reading the program was harder. Please insist if you still prefer auto.

Comment (oritwas) [reply src/rgw/rgw_admin.cc]: let's keep it as it was

File: src/rgw/rgw_admin.cc (hunk 6 above)
Comment (oritwas) [reply src/rgw/rgw_admin.cc]: we prefer to define them in source files to limit their scope

File: src/rgw/rgw_admin.cc (hunk 3 above)
Comment (oritwas) [reply src/rgw/rgw_admin.cc]: In general you should not mix changes it makes reviews easier. This should be in a separate commit.

File: src/rgw/rgw_admin.cc (hunk 5 above)
Comment (cache-nez) [reply src/rgw/rgw_admin.cc]: I fixed #includes and now newly introduced .h files don't include common/errno.h, so I think I don't.

**Comment by mattbenjamin** [comment]
@alimaredia ditto

**Comment by alimaredia** [comment]
unstale.

**Comment by tchaikov** [comment]
@alimaredia ping?
